### PR TITLE
Added the referral token

### PR DESCRIPTION
**Pull request by shunr-hpe**
With this change a unique referral token is created for each POST
and PUT of /bootparameters. This referral token is included in the
bootscript string. It can be used by a client to verify that boot
parameters and resulting bootscript are the parameters that they
previously set.

Jira: CASMHMS-4903

### Summary and Scope

Changes POST and PUT to return a UUID for each request. This id is also included in the boot script. It hopefully can be used by BOS to verify that a node is booted to the expected configuration.

### Issues and Related PRs

* Resolves CASMHMS-4903

### Testing

Tested on:

* docker compose env. Tested POST, and PUT for boot parameters. Tested GET for the bootscript
* wasp - Tested PUT and GET bootscript. For boot script verified that the old BSS version works with both the new and old etcd data. Also verified the same for the new BSS version. Plus I ran the smoke and functional tests. See the test log attached to this PR,

Were the install/upgrade based validation checks/tests run? Y
Were continuous integration tests run? Y
Was an Upgrade tested?                 Y
Was a Downgrade tested?                Y
If schema changes were part of this change, how were those handled in your upgrade/downgrade testing?

### Risks and Mitigations

low risk

### CODE REVIEW

**Comment by shunr-hpe** [comment]
log from testing on wasp:
[hms-bss-uuid-casmhms-4903.md](https://github.com/Cray-HPE/hms-bss/files/7832173/hms-bss-uuid-casmhms-4903.md)

**Review by nieuwsma — CHANGES_REQUESTED** [review]
Can you fix the license issues: https://github.com/Cray-HPE/.github/runs/4857593647?check_suite_focus=true

We either need that or an exception. Ask Mikhail for details on how to do the exception.

**Review by nieuwsma — CHANGES_REQUESTED** [review]

File: cmd/boot-script-service/boot_data.go [hunk 1]
```diff
@@ -69,10 +71,11 @@ type ImageData struct {
...
 type BootData struct {
-	Params    string
-	Kernel    ImageData
-	Initrd    ImageData
-	CloudInit bssTypes.CloudInit `json:"cloud-init,omitempty"`
+	Params        string
+	Kernel        ImageData
+	Initrd        ImageData
+	CloudInit     bssTypes.CloudInit `json:"cloud-init,omitempty"`
+	ReferralToken string             `json:"referral-token,omitempty"`
```
[the remark this answers is not in the document]
Comment (shunr-hpe) [reply cmd/boot-script-service/boot_data.go:78]: BootData is an internal data structure. I don't know why it has json annotations. I haven't found any where that it is converted to or from json.

The REST json response is defined by the struct BootParams see https://github.com/Cray-HPE/hms-bss/blob/master/pkg/bssTypes/types.go#L53-L61

The etcd json data is defined by the struct: BootDataStore

BootData is created from the BootDataStore and the kernel and initrd data. BootParams (the json http response) is sometimes created from BootData (some times it is created from other structs depending on the situation).

I agree that the referral token shouldn't be included in any of the json response bodies. And I believe that it isn't.

[the remark this answers is not in the document]
Comment (nieuwsma) [reply cmd/boot-script-service/boot_data.go:78]: the `json:"referral-token,omitempty"` tells GO how to handle this if it is json marshalled.  So if this struct is never to be json marshalled, go ahead and remove that part of it.  Just delete the `json:"referral-token,omitempty"`.

Does that make sense?

[the remark this answers is not in the document]
Comment (shunr-hpe) [reply cmd/boot-script-service/boot_data.go]: Sounds good, I've removed it from ReferralToken on BootData. I didn't remove the pre-existing annotation on CloudInit. I'm willing to do that too.
I searched through the code again and couldn't find anywhere BootData was used as json. I looked at all the Unmarshal and Marshal calls. Plus all the references to BootData. I could have overlooked something.

[the remark this answers is not in the document]
Comment (shunr-hpe) [reply cmd/boot-script-service/boot_data.go]: I've now also removed it from Cloudinit too. I figure its better with all or none (and we are going to try none).

[the remark this answers is not in the document]
Comment (nieuwsma) [reply cmd/boot-script-service/boot_data.go]: seems reasonable. Thanks

File: cmd/boot-script-service/api_test.go [hunk 2]
```diff
@@ -1,6 +1,6 @@
 // MIT License
 //
-// (C) Copyright [2021] Hewlett Packard Enterprise Development LP
+// (C) Copyright [2021-2022] Hewlett Packard Enterprise Development LP
```
Comment (nieuwsma) [inline cmd/boot-script-service/api_test.go:3]: are you changing copyright years on files that you didnt modify?
If the file has no change, it doesnt need the copyright change.   Is the license checker failing that?

Comment (shunr-hpe) [reply cmd/boot-script-service/api_test.go:3]: Yes, I changed the copyright on the files that I didn't change. Those were the files that the license checker was failing on.

Comment (nieuwsma) [reply cmd/boot-script-service/api_test.go:3]: talk to Mikhail about this, bc we arent going to just bulk update years b/c of the license checker.

**Review by nieuwsma — APPROVED** [review]

**Pull request merged**